### DISCUSSION TITLE
very long `cp` step at the end of tasks

### DISCUSSION THREAD

**Discussion by thekyz** [concourse/concourse#7466] [General]
Hi, we're trying to migrate to concourse but I've been stuck with an annoying issue at the end of build tasks: although the "task" itself (building my code) is done, the task itself stays stuck for hours (up to 4) at the end while (from the UI) not doing anything.

Looking at the logs from the worker or the web instance did not give anything but logging on the worker I saw that the only thing that was running was a `cp` process copying a volume. What is that task going and why is it taking that long? I'm using a cache to store my build data so I'm assuming this is the issue (the cache is approximately 110G).

If it is my problem, is there any way I can speed things up? I really need to cache this as otherwise the build will take quite a long time.

FYI I'm running currently with the `overlay` driver but I tried with `btrfs` and I had the same issue. I'm using `containerd` as runtime, don't know if it can have any impact. 

e.g. this has been running for nearly 3 hours:
```
2:54 cp -a /concourse-work-dir/volumes/live/8ceaa0f7-fb96-421c-5f6f-00668e7eaefe/volume/. /concourse-work-dir/volumes/init/05bbcd27-e950-4223-6466-7e680b0d29ec/volume
```

update: so digging into the code this is coming from the overlay driver so it's definitely the problem; trying again with btrfs ...

**Comment by aoldershaw** [comment]
I don't *think* the `cp` is overlay specific - it's a result of creating an "import volume" for the task cache:
https://github.com/concourse/concourse/blob/1e24cfceeefd388d8c993ab9b26c71d05642579a/atc/worker/volume.go#L162-L176

Which I believe is driver independent:
https://github.com/concourse/concourse/blob/d61508db114d3db21c66e6b7dc927c4700bd7d02/worker/baggageclaim/volume/import_strategy.go#L29-L30

The reason for creating an import volume is two-fold:
1. If the task cache already exists on a worker, we mount a copy-on-write (COW) volume of the cache in place of the original volume so that if there are multiple builds are running in parallel, the two builds won't pollute each other's state: https://github.com/concourse/concourse/blob/1e24cfceeefd388d8c993ab9b26c71d05642579a/atc/worker/worker.go#L536-L538
2. If we initialize the COW volume as the new task cache volume, we could end up with an unbounded chain of COW volumes, which I gather causes issues

That said, it's likely that 1. is a bit too aggressive in some cases - the job may always run in serial, or multiple parallel tasks can safely share the same volume without conflict. In either case, creating a COW volume is unnecessary, meaning we wouldn't need to perform the `cp` after the task. Maybe we could add extra configuration to the task cache to instruct Concourse that a COW volume is unnecessary for a particular cache, e.g.

```yaml
caches:
- path: some-path
  no_copy: true
```

or something that expresses the reason why you might not *always* skip copying the cache, e.g.

```yaml
caches:
- path: some-path
  safe_for_concurrent_use: true # this name is awful, but just an idea
```

**Reply by thekyz** [reply]
thx for the answer, I was starting to come to the same conclusion although with less knowledge of the internal workings. IIRC today it's not possible to prevent that copy/clone from happening?

**Reply by aoldershaw** [reply]
Unfortunately no, sorry. We've run into similar slowness copying task caches (but don't even have caches nearly as big as what you're working with) so I think there's definitely value in such a feature - and it shouldn't be too hard to implement.
